File: mobile/src/sync/syncMeta.ts
import {getMeta, setMeta} from '../database/index';

export const SYNC_META_KEYS = {
  lastSeenRevision: 'forma_sync:last_seen_revision',
  lastUploadAt: 'forma_sync:last_upload_at',
  lastDownloadAt: 'forma_sync:last_download_at',
  exportWatermark: 'forma_sync:export_watermark',
  deviceId: 'forma_sync:device_id',
  lastError: 'forma_sync:last_error',
  tokenExpired: 'forma_sync:token_expired',
} as const;

export async function getLastSeenRevision(): Promise<number> {
  const raw = await getMeta(SYNC_META_KEYS.lastSeenRevision);
  const n = raw != null ? Number(raw) : 0;
  return Number.isFinite(n) ? n : 0;
}

export async function setLastSeenRevision(revision: number): Promise<void> {
  await setMeta(SYNC_META_KEYS.lastSeenRevision, String(revision));
}

export async function getExportWatermark(): Promise<string | null> {
  return getMeta(SYNC_META_KEYS.exportWatermark);
}

export async function setExportWatermark(iso: string): Promise<void> {
  await setMeta(SYNC_META_KEYS.exportWatermark, iso);
}

export async function getLastUploadAt(): Promise<string | null> {
  return getMeta(SYNC_META_KEYS.lastUploadAt);
}

export async function setLastUploadAt(iso: string): Promise<void> {
  await setMeta(SYNC_META_KEYS.lastUploadAt, iso);
}

export async function getLastDownloadAt(): Promise<string | null> {
  return getMeta(SYNC_META_KEYS.lastDownloadAt);
}

export async function setLastDownloadAt(iso: string): Promise<void> {
  await setMeta(SYNC_META_KEYS.lastDownloadAt, iso);
}

export async function getFormaSyncDeviceId(): Promise<string | null> {
  return getMeta(SYNC_META_KEYS.deviceId);
}

export async function setFormaSyncDeviceId(deviceId: string): Promise<void> {
  await setMeta(SYNC_META_KEYS.deviceId, deviceId);
}

export async function getFormaSyncLastError(): Promise<string | null> {
  const raw = await getMeta(SYNC_META_KEYS.lastError);
  return raw && raw.trim() ? raw : null;
}

export async function setFormaSyncLastError(message: string | null): Promise<void> {
  if (message) {
    await setMeta(SYNC_META_KEYS.lastError, message);
  } else {
    await setMeta(SYNC_META_KEYS.lastError, '');
  }
}

export async function isFormaSyncTokenExpired(): Promise<boolean> {
  return (await getMeta(SYNC_META_KEYS.tokenExpired)) === '1';
}

export async function setFormaSyncTokenExpired(expired: boolean): Promise<void> {
  await setMeta(SYNC_META_KEYS.tokenExpired, expired ? '1' : '0');
  if (expired) {
    await setFormaSyncLastError('Токен Яндекс.Диска истёк — переподключите облако в настройках');
  }
}

File: backend/services/forma_sync/sync_meta.py
# -*- coding: utf-8 -*-
from __future__ import annotations

import uuid

from backend.database.app_meta import meta_get, meta_set

KEYS = {
    "last_seen_revision": "forma_sync:last_seen_revision",
    "last_upload_at": "forma_sync:last_upload_at",
    "last_download_at": "forma_sync:last_download_at",
    "device_id": "forma_sync:device_id",
    "last_error": "forma_sync:last_error",
    "auto_enabled": "forma_sync:auto_enabled",
}


def get_last_seen_revision() -> int:
    raw = meta_get(KEYS["last_seen_revision"])
    if not raw:
        return 0
    try:
        return int(raw)
    except ValueError:
        return 0


def set_last_seen_revision(revision: int) -> None:
    meta_set(KEYS["last_seen_revision"], str(revision))


def get_last_upload_at() -> str | None:
    raw = meta_get(KEYS["last_upload_at"])
    return raw if raw else None


def set_last_upload_at(iso: str) -> None:
    meta_set(KEYS["last_upload_at"], iso)


def get_last_download_at() -> str | None:
    raw = meta_get(KEYS["last_download_at"])
    return raw if raw else None


def set_last_download_at(iso: str) -> None:
    meta_set(KEYS["last_download_at"], iso)


def get_last_error() -> str | None:
    raw = meta_get(KEYS["last_error"])
    return raw if raw and raw.strip() else None


def set_last_error(message: str | None) -> None:
    meta_set(KEYS["last_error"], message or "")


def get_or_create_device_id() -> str:
    existing = meta_get(KEYS["device_id"])
    if existing:
        return existing
    new_id = str(uuid.uuid4())
    meta_set(KEYS["device_id"], new_id)
    return new_id


def is_auto_enabled() -> bool:
    raw = meta_get(KEYS["auto_enabled"])
    return raw != "0"


def set_auto_enabled(enabled: bool) -> None:
    meta_set(KEYS["auto_enabled"], "1" if enabled else "0")
